src/api/uploads/MultiputPart.js

```diff
@@ -26,38 +30,291 @@ class MultiputPart extends Base {
+     *
+     * @return {Object}
+     */
+    getPart = (): Object => this.data.part || {};
+
+    /**
+     * Uploads this Part via the API. Will retry on network failures.
+     *
+     * @return {void}
+     */
+    upload = (): void => {
+        if (this.isDestroyed()) {
+            return;
+        }
+
+        if (!this.sha256) {
+            throw new Error('Part SHA-256 unavailable');
+        }
+
+        if (!this.blob) {
+            throw new Error('Part blob unavailable');
+        }
+
+        const clientEventInfo = {
+            documentHidden: document.hidden,
+            digest_retries: this.numDigestRetriesPerformed,
+            timing: this.timing,
+            parts_uploading: this.getNumPartsUploading()
+        };
+
+        const headers = {
+            'Content-Type': 'application/octet-stream',
+            Digest: `sha-256=${this.sha256}`,
+            'Content-Range': `bytes ${this.offset}-${this.rangeEnd}/${this.size}`,
+            'X-Box-Client-Event-Info': JSON.stringify(clientEventInfo)
+        };
+
+        this.startTimestamp = Date.now();
+
+        this.xhr.uploadFile({
+            url: this.sessionEndpoints.upload_part,
+            data: this.blob,
+            headers,
+            method: 'PUT',
+            successHandler: this.uploadSuccessHandler,
+            errorHandler: this.uploadErrorHandler,
+            progressHandler: this.uploadProgressHandler,
+            withIdleTimeout: true,
+            idleTimeoutDuration: this.config.requestTimeoutMs
+        });
+    };
+
+    /**
+     * Handler for upload part success
+     * 
+     * @param {Object} data
+     * @return {void}
+     */
+    uploadSuccessHandler = (data: Object) => {
+        if (this.isDestroyed()) {
+            return;
+        }
+
+        this.state = PART_STATE_UPLOADED;
+        this.consoleLogFunc(() => `Upload completed: ${this.toJSON()}.`);
+        this.data = data;
+        this.blob = null;
+        this.timing.uploadTime = Date.now() - this.startTimestamp;
+
+        this.onSuccess(this);
+
+        this.uploadedBytes = this.size;
+    };
+
     /**
-     * Upload the part
-     * TODO: implement this
+     * Handler for upload part progress event
      * 
-     * @public
+     * @param {ProgressEvent} data
      * @return {void}
      */
-    upload = (): void => {};
+    uploadProgressHandler = (event: ProgressEvent) => {
+        if (this.isDestroyed()) {
+            return;
+        }
+
+        const newUploadedBytes = parseInt(event.loaded, 10);
+        const prevUploadedBytes = this.uploadedBytes;
+        this.uploadedBytes = newUploadedBytes;
+
+        this.onProgress(prevUploadedBytes, newUploadedBytes);
+    };
+
+    /**
+     * Handler for upload part error
+     * 
+     * @param {Error} error
+     * @return {void}
+     */
+    uploadErrorHandler = (error: Error) => {
+        if (this.isDestroyed()) {
+            return;
+        }
+
+        this.consoleLogFunc(
```

Comment:
ah previously this message generator function had a costly function call, so we only want to evaluate that when we can actually console log something. but since it's now removed we can probably just do this.consoleLog